cmd/ui/src/rendering/programs/edge.self.ts

```diff
@@ -0,0 +1,159 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+/**
+ * Sigma.js WebGL Renderer Edge Program
+ * =====================================
+ *
+ * Program rendering edges as thick lines but with a twist: the end of edge
+ * does not sit in the middle of target node but instead stays by some margin.
+ *
+ * This is useful when combined with arrows to draw directed edges.
+ * @module
+ */
+import CurvedEdgeProgram from './edge.curved';
+import { Coordinates, NodeDisplayData } from 'sigma/types';
+import { Attributes } from 'graphology-types';
+import { floatColor } from 'sigma/utils';
+import { bezier } from 'src/rendering/utils/bezier';
+
+const RESOLUTION = 0.02,
+    POINTS = 2 / RESOLUTION + 2,
+    ATTRIBUTES = 6,
+    STRIDE = POINTS * ATTRIBUTES;
+
+export const getControlPointsFromGroupSize = (
+    groupPosition: number,
+    radius: number,
+    center: Coordinates,
+    invertY: boolean,
+    invertX: boolean
+) => {
+    const position = groupPosition++;
+    const step = Math.PI / 2;
+
+    const theta2 = step * position - Math.PI / 2;
+    const theta1 = theta2 + step;
+
+    const x1Offset = radius * Math.cos(theta1);
+    const y1Offset = radius * Math.sin(theta1);
+
+    const x2Offset = radius * Math.cos(theta2);
+    const y2Offset = radius * Math.sin(theta2);
+
+    if (invertX && !invertY) {
+        const control2 = { x: center.x + -1 * x1Offset, y: center.y + y1Offset };
+        const control3 = { x: center.x + -1 * x2Offset, y: center.y + y2Offset };
+
+        return { control2: control2, control3: control3 };
+    }
+    if (invertY && !invertX) {
+        const control2 = { x: center.x + x1Offset, y: center.y + -1 * y1Offset };
+        const control3 = { x: center.x + x2Offset, y: center.y + -1 * y2Offset };
+
+        return { control2: control2, control3: control3 };
+    } else if (invertY && invertX) {
+        const control2 = { x: center.x + -1 * x1Offset, y: center.y + -1 * y1Offset };
+        const control3 = { x: center.x + -1 * x2Offset, y: center.y + -1 * y2Offset };
+
+        return { control2: control2, control3: control3 };
+    } else {
+        const control2 = { x: center.x + x1Offset, y: center.y + y1Offset };
+        const control3 = { x: center.x + x2Offset, y: center.y + y2Offset };
+
+        return { control2: control2, control3: control3 };
+    }
+};
+
+export default class SelfEdgeProgram extends CurvedEdgeProgram {
+    constructor(gl: WebGLRenderingContext) {
```

Comment:
I get an eslint warning here -- JS classes already create an empty constructor by default (with the `super` call for derived classes) so it is not necessary to declare your own